src/Raven.Server/Documents/TimeSeries/TimeSeriesStorage.cs

```diff
@@ -82,10 +84,54 @@ public TimeSeriesStorage(DocumentDatabase documentDatabase, Transaction tx)
             tx.CreateTree(TimeSeriesKeysSlice);
         }
 
+        public void PurgeSegments(string collection, long upto, DocumentsOperationContext context)
+        {
+            var collectionName = _documentsStorage.ExtractCollectionName(context, collection);
+            var tableName = collectionName.GetTableName(CollectionTableType.TimeSeries);
+            var table = context.Transaction.InnerTransaction.OpenTable(TimeSeriesSchema, tableName);
+
+            if (table == null || table.NumberOfEntries == 0)
+                return;
+
+            var pendingDeletion = context.Transaction.InnerTransaction.CreateTree(PendingDeletionSegments);
+            var deleted = new List<Slice>();
+            var hasMore = true;
+            
+            while (hasMore)
+            {
+                using (var it = pendingDeletion.MultiRead(collectionName.Name))
+                {
+                    if (it.Seek(Slices.BeforeAllKeys) == false)
+                        return;
+
+                    do
+                    {
+                        var etag = it.CurrentKey.CreateReader().ReadBigEndianInt64();
+                        if (etag > upto)
+                            break;
+
+                        if (table.DeleteByIndex(TimeSeriesSchema.FixedSizeIndexes[CollectionTimeSeriesEtagsSlice], etag))
+                        {
+                            deleted.Add(it.CurrentKey.Clone(context.Allocator));
+                        }
+
+                        hasMore = it.MoveNext();
+                    } while (hasMore && deleted.Count < 1000);
```

Comment:
I don't understand this code.
You are doing things in batches, but you are holding the transaction open for as long as needed.
We need to be able to do this in multiple transactions

Reply:
We do not want to put infinite number of them in `deleted` list and purging tombstones also removes all of them at once: https://github.com/ravendb/ravendb/blob/853599a811b45c9fe933606d9a04a163ef24e5b1/src/Raven.Server/Documents/DocumentsStorage.cs#L1848

We created a separate issue to optimize this: https://issues.hibernatingrhinos.com/issue/RavenDB-14487